src/DevTools/createMockNetworkLayer/index.ts

```diff
@@ -19,65 +25,137 @@ export const createMockNetworkLayer = (mockResolvers: IMocks) => {
+      if (info.fieldName in source) {
+        return inferUnionOrInterfaceType(source[info.fieldName], info)
+      }
 
-            const id = uuid()
-            idMap.set(source, id)
-            return id
-          }
+      if (info.fieldName === "__id" || info.fieldName === "id") {
+        // if relay is looking for `__id` but we only supplied `id`
+        if ("id" in source) {
+          return source.id
         }
-        complain(info, info.returnType.inspect())
-      }) as GraphQLFieldResolver<any, any>,
-      schema,
-      resolvers: {
-        FormattedNumber: () => FormattedNumber,
-        Query: Object.entries(queryData).reduce(
-          (acc, [k, v]) => ({
-            ...acc,
-            [k]: typeof v === "function" ? v : () => v,
-          }),
-          {}
-        ),
-        Mutation: Object.entries(mutationData).reduce(
-          (acc, [k, v]) => ({
-            ...acc,
-            [k]: typeof v === "function" ? v : () => v,
-          }),
-          {}
-        ),
-      },
-    })
-  )
+
+        // relay is looking for an id to denormalize the fixture in the store
+        // but we don't want to have to specify ids for all fixtures
+        // so generate one and store it in a weak map so we don't mutate
+        // the object itself
+        if (idMap.has(source)) {
+          return idMap.get(source)
+        }
+
+        const id = uuid()
+        idMap.set(source, id)
+        return id
+      }
+      complain(info, info.returnType.inspect())
+    }) as GraphQLFieldResolver<any, any>,
+    schema,
+    resolvers: {
+      FormattedNumber: () => FormattedNumber,
+      // here we map the mock fixture entries to resolver functions if they aren't
+      // already. Relay expects functions, but we want to be able to just
+      // supply plain data for syntax convenience.
+      Query: Object.entries(mockData).reduce(
+        (acc, [k, v]) => ({
+          ...acc,
+          [k]: typeof v === "function" ? v : () => v,
+        }),
+        {}
+      ),
+      Mutation: Object.entries(mockMutationResults).reduce(
+        (acc, [k, v]) => ({
+          ...acc,
+          [k]: typeof v === "function" ? v : () => v,
+        }),
+        {}
+      ),
+    },
+  })
+}
+
+// This function tries to infer the concrete type of a value that appears
+// in a position whose type is either a union or an interface
+const inferUnionOrInterfaceType = (value, info) => {
+  if (
+    !(info.returnType instanceof GraphQLUnionType) &&
+    !(info.returnType instanceof GraphQLInterfaceType)
+  ) {
+    return value
+  }
+  if (!value || value.__typename || typeof value !== "object") {
+    return value
+  }
+
+  const unionMemberTypes =
+    info.returnType instanceof GraphQLInterfaceType
+      ? info.schema._implementations[info.returnType.name]
+      : (info.returnType._types.filter(
+          type => type instanceof GraphQLObjectType
+        ) as GraphQLObjectType[])
```

Comment:
I thought I had seen a function for this in graphql-js once, but can't find it right now.

Reply:
`schema.getPossibleTypes(AbstractType)` :tada: